lm_lstm_ptb.py

```diff
@@ -0,0 +1,900 @@
+    xr = x[::-1]
+    xr_mask = x_mask[::-1]
+    n_timesteps = x.shape[0]
+    n_samples = x.shape[1]
+    embr = tparams['Wemb'][xr.flatten()]
+    embr = embr.reshape([n_timesteps, n_samples, options['dim_word']])
+    embr_shifted = tensor.zeros_like(embr)
+    embr_shifted = tensor.set_subtensor(embr_shifted[1:], embr[:-1])
+    embr = embr_shifted
+
+
+    (states_rev, _), updates_rev = get_layer(options['encoder'])[1](tparams, embr, options, prefix='encoder_r', mask=xr_mask)
+    out_lstm = get_layer('ff')[1](tparams, states_rev, options, prefix='ff_logit_lstm_r', activ='linear')
+    out_prev = get_layer('ff')[1](tparams, embr, options, prefix='ff_logit_prev_r', activ='linear')
+    out = lrelu(out_lstm + out_prev)
+
+    logit = get_layer('ff')[1](tparams, out, options, prefix='ff_logit_r',
+                               activ='linear')
+    logit_shp = logit.shape
+    probs = tensor.nnet.softmax(
+        logit.reshape([logit_shp[0]*logit_shp[1], logit_shp[2]]))
+
+    targets = xr
+    targets_mask = xr_mask
+
+    # cost
+    x_flat = targets.flatten()
+    x_flat_idx = tensor.arange(x_flat.shape[0]) * options['n_words'] + x_flat
+    cost = -tensor.log(probs.flatten()[x_flat_idx])
+    cost = cost.reshape([x.shape[0], x.shape[1]])
+    cost = (cost * targets_mask).sum(0)
+
+    return cost, states_rev, updates_rev
+
+
+    #return nll_rev, states_rev, updates_rev
+
+
+# build a training model
+def build_gen_model(tparams, options, x, y, x_mask, zmuv, states_rev):
+    n_timesteps = x.shape[0]
+    n_samples = x.shape[1]
+
+    # input
+    emb = tparams['Wemb'][x.flatten()]
+    emb = emb.reshape([n_timesteps, n_samples, options['dim_word']])
+    emb_shifted = tensor.zeros_like(emb)
+    emb_shifted = tensor.set_subtensor(emb_shifted[1:], emb[:-1])
+    emb = emb_shifted
+
+    # disconnecting reconstruction gradient from going in the backward encoder
+    rvals, updates_gen = get_layer('latent_lstm')[1](
+        tparams, state_below=emb, options=options,
+        prefix='encoder', mask=x_mask, gaussian_s=zmuv,
+        back_states=states_rev)
+
+    states_gen, kld, rec_cost_rev = rvals[0], rvals[3], rvals[4]
+    # Compute parameters of the output distribution
+    out_lstm = get_layer('ff')[1](tparams, states_gen, options, prefix='ff_logit_lstm', activ='linear')
+    out_prev = get_layer('ff')[1](tparams, emb, options, prefix='ff_logit_prev', activ='linear')
+    out = lrelu(out_lstm + out_prev)
+    logit = get_layer('ff')[1](tparams, out, options, prefix='ff_logit',
+                               activ='linear')
+    # Compute gaussian log prob of target
+    logit_shp = logit.shape
+    probs = tensor.nnet.softmax(
+        logit.reshape([logit_shp[0]*logit_shp[1], logit_shp[2]]))
+
+    # cost
+    x_flat = x.flatten()
+    x_flat_idx = tensor.arange(x_flat.shape[0]) * options['n_words'] + x_flat
+    cost = -tensor.log(probs.flatten()[x_flat_idx])
```

Comment:
Might be more efficient to use advanced indexing.
```python
probs[T.arange(x.shape[0])[:, None],
      T.arange(x.shape[1])[None, :],
      x]
```
*from memory (or it should be something like that.)